contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenMinter.sol

```diff
@@ -0,0 +1,148 @@
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@subnet-evm-contracts/interfaces/IAllowList.sol";
+import "@subnet-evm-contracts/interfaces/INativeMinter.sol";
+import "./INativeTokenMinter.sol";
+import "../../Teleporter/ITeleporterMessenger.sol";
+import "../../Teleporter/ITeleporterReceiver.sol";
+import "../../Teleporter/SafeERC20TransferFrom.sol";
+
+// Precompiled Native Minter Contract Address
+address constant MINTER_ADDRESS = 0x0200000000000000000000000000000000000001;
+// Designated Blackhole Address
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000000;
+
+contract NativeTokenMinter is ITeleporterReceiver, INativeTokenMinter, ReentrancyGuard {
+  INativeMinter private _nativeMinter = INativeMinter(MINTER_ADDRESS);
+
+  address public constant WARP_PRECOMPILE_ADDRESS =
+      0x0200000000000000000000000000000000000005;
+
+  uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 300_000; // TODO this is a placeholder
+  bytes32 public immutable currentChainID;
+  bytes32 public immutable partnerChainID;
+
+  // Used for sending an receiving Teleporter messages.
+  ITeleporterMessenger public immutable teleporterMessenger;
+
+  error InvalidTeleporterMessengerAddress();
+  error InvalidRecipientAddress();
+  error InvalidSourceChain();
+  error InvalidPartnerContractAddress();
+  error CannotBridgeTokenWithinSameChain();
+  error Unauthorized();
+  error InsufficientPayment();
+  error InsufficientAdjustedAmount(uint256 adjustedAmount, uint256 feeAmount);
+
+  // TODO we probably want to add the original token supply from this chain to the constructor.
+  constructor(address teleporterMessengerAddress, bytes32 partnerChainID_) {
+    if (teleporterMessengerAddress == address(0)) {
+        revert InvalidTeleporterMessengerAddress();
+    }
+    teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+    currentChainID = WarpMessenger(WARP_PRECOMPILE_ADDRESS)
+        .getBlockchainID();
+
+    if (partnerChainID_ == currentChainID) {
+        revert CannotBridgeTokenWithinSameChain();
+    }
+    partnerChainID = partnerChainID_;
+  }
+
+  /**
+    * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+    *
+    * Receives a Teleporter message and routes to the appropriate internal function call.
+    */
+  function receiveTeleporterMessage(
+    bytes32 nativeChainID,
+    address nativeBridgeAddress,
+    bytes calldata message
+  ) external nonReentrant() {
+
+    // Only allow the Teleporter messenger to deliver messages.
+    if (msg.sender != address(teleporterMessenger)) {
+      revert Unauthorized();
+    }
+    // Only allow messages from the partner chain.
+    if (nativeChainID != partnerChainID) {
+      revert InvalidSourceChain();
+    }
+    // Only allow the partner contract to send messages.
+    if (nativeBridgeAddress != address(this)) {
```

Comment:
Correct me if I'm wrong, but I thought we `NativeTokenMinters` communicated exclusively with `NativeTokenReceivers` and vice-versa? The addresses would be different in that case. To validate the sender, we'd want to provide that address as a constructor argument.

Reply:
We decided that the contracts would have to be deployed with the same address with the same nonce on each chain, which would give them the same address. Otherwise we have a problem with both contracts knowing the address of the other, and we're not able to use nick's method because each contract needs a different chainID passed into the constructor.

Reply:
Ahh ok I was missing that detail. That makes much more sense, thanks for clarifying.

Reply:
Is there no precompile to allow a contract to determine the id of its chain? If there is, then the contracts can be passed both chainIDs, and then use the precompile to determine which one is theirs and which one is the partner chain.

I think the current approach is fine, I'm just surprised we can't use Nick's method.

Reply:
@bernard-avalabs We could theoretically emply Nick's method by getting the chain ID from the Warp precompile. Then compare that result against the two IDs passed as constructor arguments as you suggested. That could simplify deployment, but it is nice that for both readability and succintness that in the current implementation the contracts on the native chain and the minting chain have distinct implementations.